sdk/file_sdk/XMPFileFormat/xmp_helper.cpp

```diff
@@ -21,46 +21,38 @@ XMPHelper::XMPHelper()
     if (!SXMPFiles::Initialize(kXMPFiles_IgnoreLocalText))
       throw new Exception("SXMPFiles Error");
 
-    std::string actualPrefix;
+    string actualPrefix;
     SXMPMeta::RegisterNamespace(kMsipNamespace, "msip", &actualPrefix);
   }
   catch(XMP_Error ex){
     throw new Exception(ex.GetErrMsg());
   }
 }
 
-void XMPHelper::Serialize(SXMPMeta& metadata, const vector<Tag>& tags) {
+void XMPHelper::Serialize(SXMPMeta& metadata, const vector<pair<std::string, std::string>>& propertiesToAdd, const vector<string>& keysToRemove) {
   string propertyPath;
-  vector<std::string> propertiesToDelete;
 
-  SXMPIterator iterator(metadata, kMsipNamespace, kXMP_IterJustChildren);
-  while (iterator.Next(nullptr, &propertyPath))
-    propertiesToDelete.push_back(propertyPath);
+  for(unsigned i =0; i< keysToRemove.size(); i++)
+    metadata.DeleteProperty(kMsipNamespace, keysToRemove[i].c_str());
 
-  for(unsigned i =0; i< propertiesToDelete.size(); i++)
-    metadata.DeleteProperty(kMsipNamespace, propertiesToDelete[i].c_str());
-
-  auto properties = mip::Tag::ToProperties(tags);
-
-  for(unsigned j =0; j< properties.size(); j++){
-    string value = properties[j].first;
+  for(unsigned j =0; j< propertiesToAdd.size(); j++){
+    string value = propertiesToAdd[j].first;
     if (value.empty())
       continue;

```

Comment:
why skip? we should either set empty property or delete it

Reply:
we can't set empty property and where do you want to delete it if its not set yet?

---
In reply to: [140711551](https://github.com/AzureAD/rms-sdk-for-cpp/pull/168#discussion_r140711551) [](ancestors = 140711551)

Reply:
what if you are updating an existing property?

---
In reply to: [140722592](https://github.com/AzureAD/rms-sdk-for-cpp/pull/168#discussion_r140722592) [](ancestors = 140722592,140711551)

Reply:
SetProperty add or replace, this is how xmp works